File: platform/src/types/metrics.ts
/**
 * TypeScript types matching iris/models/metrics.py ReportMetrics.
 * Used for validating CLI ingest payloads.
 */

export type CommitOrigin = "HUMAN" | "AI_ASSISTED" | "BOT";
export type ChangeIntent = "FEATURE" | "FIX" | "REFACTOR" | "CONFIG" | "UNKNOWN";

export interface OriginMetrics {
  churn_events: number;
  churn_lines_affected: number;
}

export interface StabilizationMetrics {
  files_stabilized: number;
  files_touched: number;
  stabilization_ratio: number;
}

export interface CommitShapeMetrics {
  commit_count: number;
  median_files_changed: number;
  median_total_lines: number;
  median_lines_per_file: number;
  median_directory_spread: number;
  dominant_shape: string;
}

export interface FixLatencyMetrics {
  median_latency_hours: number;
  fast_rework_pct: number;
  rework_count: number;
}

export interface CascadeMetrics {
  total_commits: number;
  cascades: number;
  cascade_rate: number;
  median_depth: number;
}

export interface DurabilityMetrics {
  lines_introduced: number;
  lines_surviving: number;
  survival_rate: number;
  median_age_days: number;
}

export interface DuplicateMetrics {
  commits_analyzed: number;
  commits_with_duplicates: number;
  duplicate_rate: number;
  total_duplicate_blocks: number;
  median_block_size: number;
}

export interface NewCodeChurnMetrics {
  files_with_new_code: number;
  files_churned_2w: number;
  files_churned_4w: number;
  churn_rate_2w: number;
  churn_rate_4w: number;
}

export interface RevertMetrics {
  reverts: number;
  revert_rate: number;
}

export interface FixTargetMetrics {
  fixes_attracted: number;
  code_share_pct: number;
  fix_share_pct: number;
  disproportionality: number;
}

export interface AcceptanceMetrics {
  total_commits: number;
  commits_in_prs: number;
  pr_rate: number;
  single_pass_rate: number;
  median_review_rounds: number;
}

export interface FunnelStage {
  stage: string;
  count: number;
  conversion: number;
}

export interface OriginFunnel {
  stages: FunnelStage[];
  overall_conversion: number;
}

export interface StabilityMapEntry {
  directory: string;
  files_touched: number;
  files_stabilized: number;
  stabilization_ratio: number;
  churn_events: number;
}

export interface ChurnFileEntry {
  file: string;
  touches: number;
  total_lines: number;
  fix_count: number;
  chain: string;
  first_touch: string;
  last_touch: string;
}

export interface ChurnCoupling {
  file_a: string;
  file_b: string;
  co_occurrences: number;
  coupling_rate: number;
}

export interface ActivityWeek {
  week_start: string;
  week_end: string;
  commits: number;
  lines_changed: number;
  intent: Record<string, number>;
  origin: Record<string, number>;
  stabilization_ratio: number;
  churn_events: number;
  prs_merged?: number;
  pr_median_ttm_hours?: number;
}

export interface ActivityPattern {
  pattern: string;
  week: string;
  description: string;
}

export interface VelocityWindow {
  start: string;
  end: string;
  commits_per_week: number;
  stabilization_ratio: number;
  churn_rate: number;
}

export interface VelocityResult {
  commits_per_week: number;
  lines_per_week: number;
  trend: string;
  trend_change_pct: number;
  durability_correlation: string;
  windows: VelocityWindow[];
}

/**
 * The complete metrics payload from the CLI.
 * Matches ReportMetrics.to_dict() output — only non-None fields are present.
 */
export interface ReportMetrics {
  // Core (always present)
  commits_total: number;
  commits_revert: number;
  revert_rate: number;
  revert_by_origin?: Record<CommitOrigin, RevertMetrics>;
  revert_by_tool?: Record<string, { reverts: number }>;
  fix_target_by_origin?: Record<CommitOrigin, FixTargetMetrics>;
  fix_target_by_tool?: Record<string, { fixes_attracted: number }>;
  churn_events: number;
  churn_lines_affected: number;
  files_touched: number;
  files_stabilized: number;
  stabilization_ratio: number;

  // Intent
  commit_intent_distribution?: Record<ChangeIntent, number>;
  churn_by_intent?: Record<ChangeIntent, OriginMetrics>;
  stabilization_by_intent?: Record<ChangeIntent, StabilizationMetrics>;
  lines_changed_by_intent?: Record<ChangeIntent, number>;

  // Origin
  ai_detection_coverage_pct?: number;
  commit_origin_distribution?: Record<CommitOrigin, number>;
  stabilization_by_origin?: Record<CommitOrigin, StabilizationMetrics>;
  churn_by_origin?: Record<CommitOrigin, OriginMetrics>;

  // Commit shape
  commit_shape_by_origin?: Record<CommitOrigin, CommitShapeMetrics>;
  commit_shape_dominant?: string;

  // Fix latency
  fix_latency_median_hours?: number;
  fix_latency_by_origin?: Record<CommitOrigin, FixLatencyMetrics>;

  // Stability map
  stability_map?: StabilityMapEntry[];

  // Correction cascades
  cascade_rate?: number;
  cascade_rate_by_origin?: Record<CommitOrigin, CascadeMetrics>;
  cascade_rate_by_tool?: Record<string, CascadeMetrics>;
  cascade_median_depth?: number;

  // Code durability
  durability_by_origin?: Record<CommitOrigin, DurabilityMetrics>;
  durability_by_tool?: Record<string, DurabilityMetrics>;
  durability_files_analyzed?: number;

  // Acceptance rate
  acceptance_by_origin?: Record<CommitOrigin, AcceptanceMetrics>;
  acceptance_by_tool?: Record<string, AcceptanceMetrics>;

  // Origin funnel
  origin_funnel?: Record<CommitOrigin, OriginFunnel>;

  // Attribution gap
  attribution_gap?: {
    flagged_commits: number;
    total_human_commits: number;
    flagged_pct: number;
    avg_loc: number;
    avg_files: number;
    avg_interval_minutes: number;
  };

  // Churn detail
  churn_top_files?: ChurnFileEntry[];
  churn_couplings?: ChurnCoupling[];

  // Activity timeline
  activity_timeline?: ActivityWeek[];
  activity_patterns?: ActivityPattern[];

  // PR lifecycle
  pr_merged_count?: number;
  pr_median_time_to_merge_hours?: number;
  pr_median_size_files?: number;
  pr_median_size_lines?: number;
  pr_review_rounds_median?: number;
  pr_single_pass_rate?: number;

  // Duplicate block detection
  duplicate_block_rate?: number;
  duplicate_block_count?: number;
  duplicate_median_block_size?: number;
  duplicate_by_origin?: Record<CommitOrigin, DuplicateMetrics>;
  duplicate_by_tool?: Record<string, DuplicateMetrics>;

  // New code churn
  new_code_churn_rate_2w?: number;
  new_code_churn_rate_4w?: number;
  new_code_churn_by_origin?: Record<CommitOrigin, NewCodeChurnMetrics>;
  new_code_churn_by_tool?: Record<string, NewCodeChurnMetrics>;

  // Velocity
  velocity?: VelocityResult;

  // Author velocity
  author_velocity?: {
    authors: Array<{
      name: string;
      high_velocity_weeks: number;
      ai_commit_pct: number;
    }>;
  };

  // Adoption timeline
  adoption_timeline?: AdoptionTimeline;
}

export type AdoptionConfidence = "clear" | "sparse" | "insufficient";

export interface AdoptionTimeline {
  first_ai_commit_date: string;        // ISO YYYY-MM-DD
  adoption_ramp_start: string;
  adoption_ramp_end: string | null;
  adoption_confidence: AdoptionConfidence;
  total_ai_commits: number;
  pre_adoption: ReportMetrics;
  post_adoption: ReportMetrics;
}

File: iris/models/metrics.py
"""Metrics data structures shared across Iris modules."""

from dataclasses import dataclass, asdict


@dataclass(frozen=True)
class ReportMetrics:
    """All metrics produced by a Iris analysis run.

    PR fields are optional — None when PR data is unavailable.
    """

    commits_total: int
    commits_revert: int
    revert_rate: float
    churn_events: int
    churn_lines_affected: int
    files_touched: int
    files_stabilized: int
    stabilization_ratio: float

    # Revert attribution (optional — None when no reverts or no origin data)
    revert_by_origin: dict[str, dict] | None = None
    revert_by_tool: dict[str, dict] | None = None

    # Fix targeting (optional — None when insufficient fix events)
    fix_target_by_origin: dict[str, dict] | None = None
    fix_target_by_tool: dict[str, dict] | None = None

    # Intent classification metrics (always populated in v0.2+)
    commit_intent_distribution: dict[str, int] | None = None
    churn_by_intent: dict[str, dict] | None = None
    stabilization_by_intent: dict[str, dict] | None = None
    lines_changed_by_intent: dict[str, int] | None = None

    # Origin classification metrics (optional — None when no non-human commits)
    ai_detection_coverage_pct: float | None = None  # % of commits with AI attribution
    commit_origin_distribution: dict[str, int] | None = None
    stabilization_by_origin: dict[str, dict] | None = None
    churn_by_origin: dict[str, dict] | None = None

    # Commit shape metrics (optional — None when no non-merge commits)
    commit_shape_by_origin: dict[str, dict] | None = None
    commit_shape_dominant: str | None = None

    # Fix latency metrics (optional — None when no rework events detected)
    fix_latency_median_hours: float | None = None
    fix_latency_by_origin: dict[str, dict] | None = None

    # Stability map (optional — None when no directories meet minimum threshold)
    stability_map: list[dict] | None = None

    # Correction cascade metrics (optional — None when insufficient data)
    cascade_rate: float | None = None
    cascade_rate_by_origin: dict[str, dict] | None = None
    cascade_rate_by_tool: dict[str, dict] | None = None
    cascade_median_depth: float | None = None

    # Code durability metrics (optional — None when git blame data unavailable)
    durability_by_origin: dict[str, dict] | None = None
    durability_by_tool: dict[str, dict] | None = None
    durability_files_analyzed: int | None = None

    # Acceptance rate metrics (optional — None when PR commit data unavailable)
    acceptance_by_origin: dict[str, dict] | None = None
    acceptance_by_tool: dict[str, dict] | None = None

    # Origin funnel (optional — None when origin data unavailable)
    origin_funnel: dict[str, dict] | None = None

    # Attribution gap (optional — None when no flagged commits)
    attribution_gap: dict | None = None

    # Churn detail (optional — None when insufficient churning files)
    churn_top_files: list[dict] | None = None
    churn_couplings: list[dict] | None = None

    # Activity timeline (optional — None when fewer than 2 weeks of data)
    activity_timeline: list[dict] | None = None
    activity_patterns: list[dict] | None = None

    # PR lifecycle metrics (optional — None when no GitHub data available)
    pr_merged_count: int | None = None
    pr_median_time_to_merge_hours: float | None = None
    # Cycle-time distribution — populated alongside pr_merged_count.
    # Together they let the platform reconstruct org-level distribution
    # by summing bucket counts across repos.
    pr_mean_time_to_merge_hours: float | None = None
    pr_p90_time_to_merge_hours: float | None = None
    pr_pct_merged_within_24h: float | None = None
    pr_cycle_time_buckets: dict[str, int] | None = None
    pr_median_size_files: int | None = None
    pr_median_size_lines: int | None = None
    pr_review_rounds_median: float | None = None
    pr_single_pass_rate: float | None = None

    # Duplicate block detection (optional — None when diff data unavailable)
    duplicate_block_rate: float | None = None
    duplicate_block_count: int | None = None
    duplicate_median_block_size: float | None = None
    duplicate_by_origin: dict[str, dict] | None = None
    duplicate_by_tool: dict[str, dict] | None = None

    # Moved code / refactoring health (optional — None when diff data unavailable)
    moved_code_pct: float | None = None
    refactoring_ratio: float | None = None
    move_by_origin: dict[str, dict] | None = None

    # Code provenance (optional — None when blame data unavailable)
    revision_age_distribution: dict[str, float] | None = None
    pct_revising_new_code: float | None = None
    pct_revising_mature_code: float | None = None
    provenance_by_origin: dict[str, dict] | None = None

    # New code churn rate (optional — None when insufficient data)
    new_code_churn_rate_2w: float | None = None
    new_code_churn_rate_4w: float | None = None
    new_code_churn_by_origin: dict[str, dict] | None = None
    new_code_churn_by_tool: dict[str, dict] | None = None

    # Operation classification (optional — None when diff data unavailable)
    operation_distribution: dict[str, float] | None = None
    operation_dominant: str | None = None
    operation_by_origin: dict[str, dict] | None = None

    # Flow Load — WIP per ISO week (optional — None when <2 buckets emerge)
    flow_load: list[dict] | None = None

    # Flow Efficiency (optional — None when no merged PR survives filters)
    flow_efficiency_median: float | None = None
    flow_efficiency_by_intent: dict[str, float] | None = None
    flow_efficiency_by_origin: dict[str, float] | None = None
    time_in_phase_median_hours: dict[str, float] | None = None
    median_time_to_first_review_hours: float | None = None
    # Count of merged PRs that survived Flow Efficiency's filters and carry a
    # phase decomposition. The honest denominator for flow coverage — NOT
    # pr_merged_count (which also counts PRs with no phase data).
    flow_pr_count: int | None = None

    # Human Review Coverage — fraction of merged PRs with genuine human review
    # (optional — None when no merged PR exists in the window). Disambiguates
    # pr_single_pass_rate: "merged in one pass" vs "no human ever looked".
    human_review_coverage_pct: float | None = None
    human_approval_coverage_pct: float | None = None
    human_review_coverage_by_intent: dict[str, float] | None = None
    human_review_coverage_by_origin_of_pr: dict[str, float] | None = None

    # Open PR Aging — snapshot of stuck inventory (optional — None when no
    # non-draft, non-bot open PR exists at coletion time).
    open_pr_count: int | None = None
    median_open_pr_age_days: float | None = None
    p90_open_pr_age_days: float | None = None
    stale_open_pr_pct: float | None = None
    very_stale_open_pr_pct: float | None = None
    abandonment_risk_pct: float | None = None
    median_open_pr_age_by_intent: dict[str, float] | None = None
    stale_open_pr_pct_by_origin: dict[str, float] | None = None

    # Merge Strategy — per-repo classification of how PRs land
    # (optional — None when no merged PR data is available). When the
    # strategy is squash/mixed, ``commit_metrics_reliable`` is False to
    # flag that per-commit metrics for this repo are approximate (squash
    # collapses N commits into 1). Strictly per-repository — no author axis.
    merge_strategy: str | None = None  # merge|squash|rebase|mixed|unknown
    merge_strategy_dominant_share: float | None = None  # 0.0–1.0
    commit_metrics_reliable: bool | None = None

    # Repository kind — CODE when a project manifest is tracked, NON_CODE for
    # documentation / issue-board repos where nothing is built or deployed.
    # Every metric below is still computed for a NON_CODE repo; the flag says
    # they describe prose, so the repo doesn't belong in a delivery comparison.
    repo_kind: str | None = None  # CODE|NON_CODE

    # DORA (real) — populated only when external Datadog events were fetched
    # for this run. None across the board when the org has no active Datadog
    # integration. ``dora_source`` is "datadog" when populated, None otherwise.
    dora_source: str | None = None
    dora_deployments_total: int | None = None
    dora_deployments_failed: int | None = None
    dora_deployments_pending_evaluation: int | None = None
    dora_incidents_total: int | None = None
    dora_cfr: float | None = None
    dora_mttr_per_deploy_seconds_median: float | None = None
    dora_mttr_per_deploy_seconds_p90: float | None = None
    dora_mttr_per_incident_seconds_median: float | None = None
    dora_mttr_per_incident_seconds_p90: float | None = None
    dora_rollback_rate: float | None = None
    dora_rollbacks_total: int | None = None
    dora_lead_time_seconds_median: float | None = None
    dora_deploy_frequency_per_day: float | None = None
    dora_remediation_distribution: dict[str, int] | None = None

    # DORA by code origin — populated only when the analysis run had both
    # external DORA events AND local commit-origin classification (i.e. the
    # CLI ran on a real repo and fetched external events). The platform's
    # AI-vs-human correlation card (dashboard) keys off these.
    dora_cfr_by_origin: dict[str, dict] | None = None
    dora_rollback_rate_by_origin: dict[str, dict] | None = None
    # Org-wide attribution coverage for the by-origin breakdowns above:
    # known_origin_commits / total_referenced_commits. Low values mean the
    # window pulled deploys whose commits are older than the local commit
    # window — the by-origin numbers stay correct but represent a subset.
    dora_cfr_by_origin_coverage_pct: float | None = None

    def to_dict(self) -> dict:
        d = asdict(self)
        # Exclude None fields for backward compatibility
        return {k: v for k, v in d.items() if v is not None}
